bin/bank/pycbc_brute_bank

```diff
@@ -247,20 +249,52 @@ class TriangleBank(object):
     def check_params(self, gen, params, threshold):
         num_tried = 0
         num_added = 0
-        for i in range(len(tuple(params.values())[0])):
-            num_tried += 1.0
-
-            try:
-                hp = gen.generate(**{key:params[key][i] for key in params})
-            except Exception as err:
-                print(err)
-                continue
-
-            hp.gen = gen
-            hp.threshold = threshold
-            if hp not in self:
-                num_added += 1
-                self.insert(hp)
+        total_num = len(tuple(params.values())[0])
+
+        if args.num_cores == 1:
+            for i in range(total_num):
+                num_tried += 1.0
+
+                try:
+                    hp = gen.generate(**{key:params[key][i] for key in params})
+                except Exception as err:
+                    print(err)
+                    continue
+
+                hp.gen = gen
+                hp.threshold = threshold
+                if hp not in self:
+                    num_added += 1
+                    self.insert(hp)
+
+        # use multiprocessing
+        elif args.num_cores > 1:
+            waveform_cache = []
+            with multiprocessing.Pool(args.num_cores) as pool:
+                for return_wf in pool.imap_unordered(
+                    wf_wrapper, 
+                    ({k: params[k][idx] for k in params} for idx in range(total_num))
+                ):
+                    waveform_cache += [return_wf]
+            
+            for i in range(total_num):
```

Comment:
nitpic:

```
for hp in waveform_cache:

     if hp is not None:
           hp.gen = gen
           hp.threshold = threshold 
           # ...
     else:
           logging.info("Waveform generation failed!")
```

Reply:
Technically, you can do this check for each waveform inside the `for` loop `for return_wf in pool.imap_unordered`, instead of storing all of them inside waveform_cache. Also, since the order is not preserved, the 2 sections (multiprocessing and sequential) are not equivalent as adding a WF to self depends on the order.

Reply:
@raffienficiaud corrected for a cleaner code (`if hp is not None` ...).  

I didn't do as what you suggested in the second comment is actually because the determination of inserting `hp` or not has to be sequential. E.g., once the first `hp` is inserted, the second `hp` would depend how close it is to the first `hp`. There could be a way to further optimize this, i.e., only preserve those `hp` not close to each other, then throw them into the multiprocessing. But I'll be lazy to not consider it in this PR

Reply:
> actually because the determination of inserting hp or not has to be sequential

This is actually my point: the order matters but it is lost in the call to `imap_unordered` as this does not preserve the order. I believe this should not be an issue, except if you want reproducibility. You can easily re-order the elements if you pass the index of that element to the wrapper, and retrieve that index back in the return statement.

Reply:
@raffienficiaud The reproducibility is indeed lost. I presume it's not a problem in this particular project so not to implement this feature. But I'd be interested in learning this order restoring trick though! I'd write the following 
```
def wf_wrapper(p, index):
    try:
        hp = gen.generate(**p)
        return hp, index
    except Exception:
        return None, index
```
then I'm not quite sure what's an easy way to pass the index into the wrapper, probably as follows?
```
        for return_wf in pool.imap_unordered(
                wf_wrapper, 
                ({k: params[k][idx] for k in params}, idx for idx in range(total_num))
            ):
......
```

Reply:
Yes, something along those lines.